axelrod/match.py

```diff
@@ -35,8 +40,17 @@ def __init__(self, players, turns, game=None, deterministic_cache=None,
             The default is to use the correct values for turns, game and noise
             but these can be overridden if desired.
         """
-        self.result = []
+
+        if prob_end is not None:
+            if turns is not None:
+                turns = min(turns, sample_length(prob_end))
+            else:
+                turns = sample_length(prob_end)
+        elif turns is None:
+            turns = 20
```

Comment:
I know it's my usual reaction to conditional statement but, especially with nesting, I'd be tempted to use a dict here. Something like:

```
turns_mapping = {
    (True, True): 20,
    (True: False): sample_length(prob_end),
    (False, True): turns,
    (False, False): min(turns, sample_length(prob_end))
}

turns = turns_mapping[(turns is None, prob_end is None)]
```